### PR TITLE
Add multi recipe feedstock

### PR DESCRIPTION
**Pull request by janschulz**
This adds a single switch to init (and a config file value), which  makes this feedstock take multiple recipes in subdirs under ./recipes  (the single case has the recipe directly in ./recipe).
    
All packages are shown in the generated reade, so when you add packages, you need call "conda smithy regenerate" to get the overview of all  packages...

This PR is based on (and therefore includes) https://github.com/conda-forge/conda-smithy/pull/56, so that should be merged first...

replaces https://github.com/conda-forge/conda-smithy/pull/51

Closes: #37 the issue asking for multi recipe support
Closes: #51 the old PR

### CODE REVIEW

**Comment by JanSchulz** [comment]
Rebased on top of #56, only the last commit is new (and interesting...)

File: conda_smithy/ci_register.py [hunk 1]
```diff
@@ -15,20 +16,34 @@
...
 try:
     with open(os.path.expanduser('~/.conda-smithy/appveyor.token'), 'r') as fh:
         appveyor_token = fh.read().strip()
 except IOError:
-    print('No appveyor token. Put one in ~/.conda-smithy/appveyor.token')
+    print('No appveyor token. Create a token at https://ci.appveyor.com/api-token and\n'
+          'put it in ~/.conda-smithy/appveyor.token')
+
+try:
+    anaconda_token = os.environ['BINSTAR_TOKEN']
```
Comment (pelson) [inline conda_smithy/ci_register.py]: I'm completely game for changing APIs. We are still in 0.x versions. I think it makes sense just to ditch this try..except altogether.

File: conda_smithy/ci_register.py [hunk 2]
```diff
@@ -149,7 +162,10 @@ def add_project_to_travis(user, project):
 
         if count > 5:
             print('  * Unable to register the repo on Travis (is it down?)')
-            return
+            print("    Please also ensure that the user/org name is spelled correctly "
+                  "(case sensitiv).")
+
+            sys.exit(1)
```
Comment (pelson) [inline conda_smithy/ci_register.py]: :-1: on that. Is there a reason you ```sys.exit```ed? I really don't want conda-forge to be a CLI only.

File: conda_smithy/ci_register.py [hunk 3]
```diff
@@ -149,7 +162,10 @@ def add_project_to_travis(user, project):
 
         if count > 5:
             print('  * Unable to register the repo on Travis (is it down?)')
```
Comment (pelson) [inline conda_smithy/ci_register.py]: Perhaps just add ``'(is it down? is the {} slug good [note: case sensitive])?'.format(repo_slug)``

File: conda_smithy/cli.py [hunk 4]
```diff
@@ -12,39 +12,83 @@
 from . import configure_feedstock
 from . import lint_recipe
 
+_CONFIG_CI_PLACEHOLDER = '<will be updated by "conda-smithy register-ci">'
+_CONFIG_GH_PLACEHOLDER = '<will be updated by "conda-smithy register-github">'
+
+PY2 = sys.version_info[0] == 2
```
Comment (pelson) [inline conda_smithy/cli.py]: six is a reasonable dependency to add.

File: conda_smithy/cli.py [hunk 5]
```diff
@@ -54,62 +98,87 @@ class Init(Subcommand):
...
-        subcommand_parser.add_argument("--feedstock-directory",
-                                       default='./{package.name}-feedstock')
-        subcommand_parser.add_argument("--no-git-repo", action='store_true',
-                                       default=False)
+
+        super(Init, self).__init__(parser, "Create a feedstock git repository, which can contain "
+                                           "one or multiple conda recipes.")
+        scp = self.subcommand_parser
+        scp.add_argument("recipe_directory", help="The path to the source recipe directory.")
+        scp.add_argument("--multi", action='store_true', default=False,
```
Comment (pelson) [inline conda_smithy/cli.py]: I thought the multi part was going to be taken out of this PR?

**Comment by pelson** [comment]
Just to re-iterate - there are some great changes in here - but the PR is just too broad in scope. Things like pulling out the use of BINSTAR_TOKEN will be hugely valuable, and I'm keen for them to be done in small PRs where possible (in this case, I believe there are several completely independent improvements).

File: conda_smithy/ci_register.py (hunk 2 above)
Comment (JanSchulz) [inline conda_smithy/ci_register.py]: ok, then this needs something different: The problem was that `gegister-ci` calls this methods and then all otheres and does not check if the first works and so the second call to travis related stuff will also fail:

```
        ci_register.add_project_to_travis(owner, repo) # -> doe snot work, but returns ok
        ci_register.travis_token_update_conda_forge_config(args.feedstock_directory, owner, repo) # -> fails as well
```

So I think the right action would be to add a `UnrecoverableException(msg)` and do a 

```
try:
   ... real work
except UnrecoverableException as a:
    print(a)
    return
```

in all cli functions...

File: conda_smithy/cli.py (hunk 5 above)
Comment (JanSchulz) [inline conda_smithy/cli.py]: you are reviewing the multi PR (which includes the other PR) :-)

-> You probably want to start with https://github.com/conda-forge/conda-smithy/pull/56

File: conda_smithy/ci_register.py (hunk 2 above)
Comment (pelson) [inline conda_smithy/ci_register.py]: Good point. Simply raising something useful sounds sensible to me. To persist the behaviour the CLI can catch that and continue beyond Travis things if we like.

File: conda_smithy/ci_register.py (hunk 1 above)
Comment (JanSchulz) [inline conda_smithy/ci_register.py]: ok, done

File: conda_smithy/ci_register.py (hunk 3 above)
Comment (JanSchulz) [inline conda_smithy/ci_register.py]: changed to 
```
msg = ('  * Unable to register the repo on Travis\n'
                   '    (Is it down? Is the "{}" name spelled right [note: case sensitive]?)')
            raise RuntimeError(msg.format(user))
```

File: conda_smithy/ci_register.py (hunk 2 above)
Comment (JanSchulz) [inline conda_smithy/ci_register.py]: Use `RuntimeError`

Comment (JanSchulz) [inline conda_smithy/ci_register.py]: BTW: should the cli commands (e.g. all `__call__`) return or use `sys.exit(1)` if an error occurred?

**Comment by pelson** [comment]
Many of the excellent ideas from this PR were pulled in. The major blocking point was the many-recipes.
Since this has now gone stale, and we have reached the point of ruling out multiple-recipe feedstocks for the time being, I'm going to close this.

Thanks for taking the time to put this together @janschulz, and in particular for splitting off many of the parts into separate PRs.

**Pull request closed without merging**